2562 - Per-file parsing log, tech memo (#3275)
raft-tech/TANF-app: docs/Technical-Documentation/tech-memos/parsing-log-per-file/parsing-log-per-file.md

# Parsing log per file upload

**Audience**: TDP Software Engineers <br>
**Subject**:  Parsing log per file upload <br>
**Date**:     November 4, 2024 <br>

## Summary
This technical memorandum discusses the implementation of features to bring more visibility into system behavior during file parsing. This includes:

> * Generating a storing a file for logs generated during the parsing run. The log file should be stored in s3, alongside the submitted datafile.

Reviewer: ```suggestion
* Generating and storing a file for logs generated during the parsing run. The log file should be stored in s3, alongside the submitted datafile.
```

> This memo discussed various proposed implementations and the benefits/drawbacks of each, then provides a suggested implementation taking these factors into consideration.

Reviewer: ```suggestion
This memo discusses various implementations and the benefits/drawbacks of each, then provides a suggested implementation taking these factors into consideration.
```

Author: yeesh. reasons why i didn't major in english 😄

* These logs possibly leak sensitive data
* Hard to dig back through logs to file errors associated with a particular upload
* No visibility into differences in logs between parse runs for the same file

## Out of Scope
Call out what is out of scope for this technical memorandum and should be considered in a different technical memorandum.
* Efficiency - must write logs to file on disk and upload to s3 at the end of the parser run.
   * This will have a memory impact and a disk space impact, as well as increase the run time of the parsing process (network upload).
   * The singleton solution explored here will additionally increase memory utilization.
   * Some mitigation techniques are mentioned, but not implemented.
* Association of the parse log file with a model in the Django Admin Console.
   * An MVP of this feature only includes uploading the resulting file to s3, alongside the datafile submission.
   * An exploration of this was done in Method/Design Step 6, but is out of scope for this work.

## Method/Design

In general, this solution requires two simple parts:

> 1. Captures the log output during parsing of a datafile and write it to a file on disk

Reviewer: ```suggestion
1. Capture the log output during parsing of a datafile and write it to a file on disk
```

   * This solution utilizes the `FileHandler` already included as part of python's `logging` utility, which we use extensively throughout TDP (including throughout the parser). `FileHandler` allows log output to be written to a file on disk
      * [Documentation](https://docs.python.org/3/howto/logging.html#logging-to-a-file)
      * Example:
         ```python
         import logging

         def setup_logger(name, filename):
            handler = logging.FileHandler(filename)
            logger = logging.getLogger(name)
            logger.addHandler(handler)
            logger.setLevel(logging.INFO)
            return logger, handler

         logger, handler = setup_logger('test123', 'test123.log')
         logger.info('test123')
         logger.warning('asdddddddd3')
         logger.error('asdfasdf')
         handler.close()
         ```
   * This step can be accomplished using a custom-built file logger. However, since `logging` is already used extensively throughout the project, we chose to extend this existing functionality rather than have to replace it for a single use-case.

>    * This comes with the drawback that `logging.FileHandler` doesn't compress the resulting file by default. We could potentially address this by overriding or creating a custom version of `logging.FileHandler`.

Reviewer: Thats a good idea!

> 2. Upload the resulting file to s3 once parsing completes

Reviewer: I was thinking it could be nice to call out why S3 is "better" as opposed to storing the file in postgres. I caught myself asking the question again and then reasoning about it while reading. Your call though, im good either way!

Reviewer 2: I want to also call out that although s3 is nice option, there is still 5TB size limit (it's a lot but can be filled out quickly), we might need to think about retention policy. S3 has different tiers, a less frequent access tier is: S3 Standard-IA we might want to look into that to backup logs after some time.

Author: @elipe17 @raftmsohani updated, let me know what you think. i couldn't think of a reason we'd put the files in postgres, so i didn't mention alternatives to s3. just some pros and cons

   * Once parsing completes and a file containing logs is written to disk, we can use existing tooling in our application to upload the resulting file to s3. An example of this is already implemented in `tdpservice.search_indexes.tasks.export_queryset_to_s3_csv`
   * Example
      ```python
      from botocore.exceptions import ClientError
      from tdpservice.data_files.s3_client import S3Client

      s3 = S3Client()
      try:
         s3.client.upload_file(local_filename, settings.AWS_S3_DATAFILES_BUCKET_NAME, s3_file_path)
      except ClientError as e:
         logger.error('upload to s3 failed')
      ```
   * This simply uploads a local file to s3, using the `boto` library.

>    * To limit disk usage, we could potentially write logs to s3 on-the-fly as parsing happens using the `smart_open` library.

Reviewer 2: this might cause responsiveness issues

Author: Right after this line, i mention 
> This would incur additional network timing/cost as parsing runs, as logs would need to be transmitted to s3 many times during the parse execution, rather than once after.

is there another responsiveness issue you're referring to? or does this capture your concerns?

   ```
   * This requires a lot of passing of variables, creating the opportunity to miss a module or function
* Create a global (singleton) class that holds on to instances of `logging` so that other functions can access them.
   ```python
   class ParseLogContextManager:
      """Caches an instance of logger for use throughout the parse routine."""

      def __init__(self, logger_instance):
         self.logger_instance = None
         self.handler_instance = None

      def set_logger_instance(self, logger_name):
         """Creates and caches a new instance of logger."""
         logger, handler = setup_logger(logger_name, f'{logger_name}.log')
         self.logger_instance = logger
         self.handler_instance = handler

      def get_logger_instance(self):
         """Returns the cached instance of logger."""
         return self.logger_instance

      def clear_logger_instance(self):
         """Closes and clears the parse logger instance."""
         self.handler_instance.close()
         # upload to s3
   ```
   * This is easier to implement as it doesn't require modifying every function call in the parser

>    * This likely has a higher memory impact, though, as an instance of this class will be present in memory for every instance of the application

Reviewer 2: How much memory does it need? I can't imagine more a couple of KB

Author: agreed, it shouldn't take much. there is an impact though, and i think it would be correlated to how many instances of `logger` are being stored

```python
from tdpservice.parsers.parse_log_context_manager import ParseLogContextManager

PARSE_LOG_CONTEXT_MANAGER = ParseLogContextManager()
```

### 2. Initialize a new logger in `parser_task`

In `tdpservice.scheduling.parser_task`, import settings
```python
from django.conf import settings
```

Then create a new logger instance at the beginning of the `parse` function
```python
settings.PARSE_LOG_CONTEXT_MANAGER.create_logger_instance(datafile_id, reparse_id)
```

At the end of the `parse` function, close the instance
```python
settings.PARSE_LOG_CONTEXT_MANAGER.clear_logger_instance(datafile_id, reparse_id)
```

### 3. Implement the remainder of `ParseLogContextManager`

Now, in `tdpservice.parsers.parse_log_context_manager`, we need to implement `ParseLogContextManager`'s remaining methods
```python
class ParseLogContextManager:
   """Caches an instance of logger for use throughout the parse routine."""

   def __init__(self, logger_instance):
      self.loggers = {}

   # can utilize this to avoid having to pass both params every time
   # (also accept `logger_name` as the param in functions below)

>    def _get_logger_name(self, datafile_id, reparse_id):

Reviewer: what happens if `reparse_id` is `None`? Do we have to pass the `reparse_id` to all parse functions to be able to call `get_logger_instance` or is it available in the datafile model?

Author: this part takes a little finagling. We need some unique information to recall the logger from the manager class - in my mind, `datafile_id` is *probably* enough, but i included `reparse_id` to differentiate the resulting log files. 

either way, enough information to reconstruct the identifier (or the identifier itself) needs to be passed into functions that require the logger. I don't see a good way around that

Reviewer: I think so too. Ill do some more thinking about how we can avoid passing the ID(s) and see if I come up with anything. Non blocking though!

* `tdpservice.scheduling.parser_task`
* `tdpservice.parsers.parse`
* `tdpservice.parsers.case_consistency_validator`
* `tdpservice.parsers.duplicate_manager`
* `tdpservice.parsers.fields`
* `tdpservice.parsers.row_schema`
* `tdpservice.parsers.util`

### 5. Implement the s3 upload

The s3 upload can be implemented with the following simple code:
```python
from botocore.exceptions import ClientError
from tdpservice.data_files.s3_client import S3Client

s3 = S3Client()
try:
   s3.client.upload_file(local_filename, settings.AWS_S3_DATAFILES_BUCKET_NAME, s3_file_path)
except ClientError as e:
   logger.error('upload to s3 failed')
```

This just needs to be placed where it makes the most sense. Make sure `handler.close()` has been called (such as is done in `ParseLogContextManager.clear_logger_context`).

### 6. (Optional) Associate logs to a submission in the Django Admin Console

To associate the uploaded log file with a submission in the admin console, we can
1. Add the s3 url to the existing `DataFile` or `DataFileSummary` models

>    * These get wiped out with every reparse, so we would be unable to see log differences between parses of the same submission.

Reviewer: Is there a reason we couldnt have a one to many relation on the datafile? The datafile model is no longer deleted during a reparse.

Author: Having a url property on `DataFile` is what I was referring to being wiped out. We could absolutely do a one-to-many relationship by creating another model to hold the s3 url

Reviewer: Right my brain wasn't working. For some reason I was thinking we could have a foreign key on the URL but that doesnt make sense. Thank you.
